File: src/shared/agentOrder.ts
import type { AgentId, AgentInfo } from './transcript';

// acpira.agentOrder / acpira.disabledAgents: the order every agent list follows and the agents kept out of the new-session entry points
export interface AgentPrefs {
  order: AgentId[];
  disabled: AgentId[];
}

// Listed ids first in their saved order, the rest (new built-ins, fresh custom agents) keep registry order after them.
// External entries stay last. Disabled agents are flagged, never dropped: history, settings and running sessions still need them
export function arrangeAgents<A extends AgentInfo>(list: A[], prefs: AgentPrefs): A[] {
  const rank = new Map(prefs.order.map((id, i) => [id, i]));
  const off = new Set(prefs.disabled);
  const key = (a: A, i: number) => [a.external ? 1 : 0, rank.get(a.id) ?? prefs.order.length, i] as const;
  return list
    .map((a, i) => ({ a: off.has(a.id) && !a.external ? { ...a, disabled: true } : a, k: key(a, i) }))
    .sort((x, y) => x.k[0] - y.k[0] || x.k[1] - y.k[1] || x.k[2] - y.k[2])
    .map(x => x.a);
}

// Agents a new-session entry point offers (the plus menu, the default agent picker)
export const launchable = (agents: AgentInfo[]) => agents.filter(a => !a.external && !a.disabled);

// The configured default unless it was switched off; then the first enabled agent, preferring an installed one.
// An id the list does not know stays as configured (a custom agent not probed yet)
export function pickDefaultAgent(agents: AgentInfo[], preferred: AgentId): AgentId {
  if (!agents.find(a => a.id === preferred)?.disabled) return preferred;
  const open = launchable(agents);
  return (open.find(a => a.available !== false) ?? open[0])?.id ?? preferred;
}

// Move one id to another position of the full visible list; the result is what agentOrder stores
export function moveAgent(ids: AgentId[], id: AgentId, to: number): AgentId[] {
  const from = ids.indexOf(id);
  if (from < 0) return ids;
  const next = ids.filter(x => x !== id);
  next.splice(Math.max(0, Math.min(next.length, to)), 0, id);
  return next;
}

File: rust/crates/acpira-shared/src/agent_order.rs
//! Agent ordering and enablement (mirror of src/shared/agentOrder.ts)

use std::collections::{HashMap, HashSet};

use crate::transcript::{AgentId, AgentInfo};

#[derive(Debug, Clone, Default)]
pub struct AgentPrefs {
  pub order: Vec<AgentId>,
  pub disabled: Vec<AgentId>,
}

/// Listed ids first in their saved order, the rest keep registry order after them; external entries stay last.
/// Disabled agents are flagged, never dropped
pub fn arrange_agents(list: Vec<AgentInfo>, prefs: &AgentPrefs) -> Vec<AgentInfo> {
  let rank: HashMap<&str, usize> = prefs.order.iter().enumerate().map(|(i, id)| (id.as_str(), i)).collect();
  let off: HashSet<&str> = prefs.disabled.iter().map(String::as_str).collect();
  let mut keyed: Vec<((u8, usize, usize), AgentInfo)> = list
    .into_iter()
    .enumerate()
    .map(|(i, mut a)| {
      let external = a.external == Some(true);
      let k = (external as u8, rank.get(a.id.as_str()).copied().unwrap_or(prefs.order.len()), i);
      if off.contains(a.id.as_str()) && !external {
        a.disabled = Some(true);
      }
      (k, a)
    })
    .collect();
  keyed.sort_by_key(|(k, _)| *k);
  keyed.into_iter().map(|(_, a)| a).collect()
}
